docs/src/internals.md

```diff
@@ -0,0 +1,256 @@
+# Internals and extending the formula DSL
+
+This section is intended to help **package developers** understand the internals
+of how a `@formula` becomes a numerical matrix, in order to use, manipulate, and
+even extend the DSL.  The Julia `@formula` is designed to be as extensible as
+possible through the normal Julian mechanisms of multiple dispatch.
+
+## The lifecycle of a `@formula`
+
+A formula goes through a number of stages, starting as an
+expression that's passed to the `@formula` macro and ending up generating a
+numeric matrix when ultimately combined with a tabular data source:
+
+1. "Macro time" when only the surface syntax is available
```

Comment:
```suggestion
1. "Macro time" when only the surface syntax is available.
```
Maybe note that this can happen at runtime if you pass a generated string?